Src/main.c

```diff
@@ -102,12 +102,40 @@ int main(void)
 
 
   HAL_GPIO_WritePin(LED1_GPIO_Port, LED1_Pin, GPIO_PIN_SET);
+
+  if(HAL_ADC_Start(&hadc) != HAL_OK)
+  {
+    _Error_Handler(__FILE__, __LINE__);
+  }
+  if(HAL_TIM_PWM_Start(&htim16, TIM_CHANNEL_1) != HAL_OK)
+  {
+    _Error_Handler(__FILE__, __LINE__);
+  }
+  const uint32_t PWM_COMPARE_MIN = 3199; //1 millisecond duty cycle
+  const uint32_t PWM_COMPARE_MAX = 6399; //2 millisecond duty cycle
+  const uint32_t PWM_COMPARE_DIFF = PWM_COMPARE_MAX - PWM_COMPARE_MIN;
+  __HAL_TIM_SET_COMPARE(&htim16, TIM_CHANNEL_1, PWM_COMPARE_MIN);
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
   while (1)
   {
+    const int TIMEOUT_MILLIS = 1000000;
+    uint32_t adcValue;
+    if(HAL_ADC_PollForConversion(&hadc, TIMEOUT_MILLIS) == HAL_OK)
+    {
+      adcValue = HAL_ADC_GetValue(&hadc);
+    }
+    else
+    {
+      _Error_Handler(__FILE__, __LINE__);
+    }
+
+    const uint32_t ADC_VAL_MAX = 4095; //For 12-bit ADC resolution
+    //Map ADC value to the range of desired PWM compare register values
+    uint32_t pwmCompare = adcValue * PWM_COMPARE_DIFF / ADC_VAL_MAX + PWM_COMPARE_MIN; //Error of +0 -1, or +0ms -3.125E-4ms due to integer division
```

Comment:
This is something that I actually missed on my first attempt too.

If adcValue is 5, pwmCompare should be 3203, however, your program outputs 3202. 

This has to do with truncation, which takes place during integer division. Looking at [this link](https://www.quora.com/In-C-integer-division-truncates-Please-explain-this-line#:~:text=When%20integers%20are%20divided%20in,is%20ceiling%20of%20the%20output%20) should help you understand truncation better.

To solve this, you can look into different datatypes and see what happens to decimal places when you divide them. [This link should get you started](https://stackoverflow.com/questions/60088323/how-i-can-prevent-integers-from-truncating-value-while-division-in-c)